zokrates_analysis/src/reducer/mod.rs

```diff
@@ -325,74 +275,70 @@ impl<'ast, 'a, T: Field> ResultFolder<'ast, T> for Reducer<'ast, 'a, T> {
-                        self.complete &= !blocked;
+                let to = self.ssa.fold_uint_expression(to);
+                let to = self.propagator.fold_uint_expression(to)?;
+                let to = self.fold_uint_expression(to)?;
+                let to = self.propagator.fold_uint_expression(to)?;
 
-                        Ok(out_statements)
-                    }
-                    _ => {
-                        let from = self.fold_uint_expression(from)?;
-                        let to = self.fold_uint_expression(to)?;
-                        self.complete = false;
-                        self.for_loop_versions_after.push(versions_before);
-                        Ok(vec![TypedStatement::For(v, from, to, statements)])
+                match (from.as_inner(), to.as_inner()) {
+                    (UExpressionInner::Value(from), UExpressionInner::Value(to))
+                        if to - from > MAX_FOR_LOOP_SIZE =>
+                    {
+                        Err(Error::LoopTooLarge(to.saturating_sub(*from)))
                     }
-                }
+                    (UExpressionInner::Value(from), UExpressionInner::Value(to)) => Ok((*from
+                        ..*to)
+                        .flat_map(|index| {
+                            std::iter::once(TypedStatement::definition(
+                                v.clone().into(),
+                                UExpression::from(index as u32).into(),
+                            ))
+                            .chain(statements.clone())
+                            .map(|s| self.fold_statement(s))
+                            .collect::<Vec<_>>()
+                        })
+                        .collect::<Result<Vec<_>, _>>()?
+                        .into_iter()
+                        .flatten()
+                        .collect::<Vec<_>>()),
+                    _ => Err(Error::NonConstant(format!(
+                        "Expected loop bounds to be constant, found {}..{}",
+                        from, to
+                    ))),
+                }?
+            }
+            s => {
+                let statements = self.ssa.fold_statement(s);
+
+                let statements = statements
+                    .into_iter()
+                    .map(|s| self.propagator.fold_statement(s))
+                    .collect::<Result<Vec<_>, _>>()?
+                    .into_iter()
+                    .flatten();
+
+                let statements = statements
+                    .map(|s| fold_statement(self, s))
+                    .collect::<Result<Vec<_>, _>>()?
+                    .into_iter()
+                    .flatten();
+
+                let statements = statements
+                    .map(|s| self.propagator.fold_statement(s))
+                    .collect::<Result<Vec<_>, _>>()?
+                    .into_iter()
+                    .flatten();
+
+                statements.collect()
```

Comment:
Just checking, is it possible to collect only once here?

Reply:
We collect a single time into `Vec`, but we need more collects into `Result<Vec<>, _>` for error handling. I had a look and did not find another easy way to do `vec.into_iter().collect::<Result<Vec<_>, _>>().flatten()` which does not require going through `Vec<_>`. This is something we could improve across the codebase if we have proof that it actually makes a difference and isn't optimised by rustc already. I don't think it's necessary in this PR.